### PR TITLE
[TwigComponent] Add documentation about passing blocks to embedded components

### PR DESCRIPTION
**Pull request by sneakyvv**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Tickets       | Fix #...
| License       | MIT

This adds documentation of the feature added in #920

### CODE REVIEW

File: src/TwigComponent/doc/index.rst [hunk 1]
```diff
@@ -832,6 +832,140 @@ The ``with`` data is what's mounted on the component object.
 
     Embedded components *cannot* currently be used with LiveComponents.
 
+Passing blocks to embedded components
```
Comment (weaverryan) [inline src/TwigComponent/doc/index.rst]: Uppercase Blocks, Embedded and Components

File: src/TwigComponent/doc/index.rst [hunk 2]
```diff
@@ -832,6 +832,140 @@ The ``with`` data is what's mounted on the component object.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 2.10
+
+    The `outerBlocks` variable has been added since 2.10.
```
Comment (weaverryan) [inline src/TwigComponent/doc/index.rst]:
```suggestion
    The ``outerBlocks`` variable was added in 2.10.
```

File: src/TwigComponent/doc/index.rst [hunk 3]
```diff
@@ -832,6 +832,140 @@ The ``with`` data is what's mounted on the component object.
...
+
+.. code-block:: twig
+
+    {# templates/successAlert.html.twig #}
+    {% component Alert with { type: 'success' } %}
+        I can override the content, but that's pretty dull.
+    {% endcomponent %}
+
+What if this SuccessAlert component wants to make the content of its nested Alert component configurable?
+Well, you can use the special `outerBlocks` variable for those cases. This makes it possible to refer
```
Comment (weaverryan) [inline src/TwigComponent/doc/index.rst]: In rst, unlike markdown, all technical words have 2 “ticks” around them, not just 1

File: src/TwigComponent/doc/index.rst [hunk 4]
```diff
@@ -832,6 +832,140 @@ The ``with`` data is what's mounted on the component object.
...
+    {% endcomponent %}
+
+.. code-block:: twig
+
+    {# templates/some_page.html.twig #}
+    {% component SuccessAlert %}
+        Look mom, I'm a complex component!
+    {% endcomponent %}
+
+Note that to pass a block multiple components down, each component would need to pass it along.
```
Comment (weaverryan) [inline src/TwigComponent/doc/index.rst]: I’m not sure about this part - it seems obvious enough to me, and the code below REALLY makes this look complex. I vote for mentioning this perhaps, but not showing it.

File: src/TwigComponent/doc/index.rst [hunk 5]
```diff
@@ -832,6 +832,140 @@ The ``with`` data is what's mounted on the component object.
...
+            {{ blocks(outerBlocks.foo) }}
+        {% endblock %}
+    {% endcomponent %}
+
+.. code-block:: html+twig
+
+    {# templates/target.html.twig #}
+    <div>{% block foo %}default content for foo{% endblock %}</div>
+
+Context of embedded components
```
Comment (weaverryan) [inline src/TwigComponent/doc/index.rst]: Uppercase everything except of

File: src/TwigComponent/doc/index.rst [hunk 6]
```diff
@@ -832,6 +832,140 @@ The ``with`` data is what's mounted on the component object.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Every embedded component can be viewed as another template inside a template, hence the name.
+Or in other words, in the other direction, it's as if the embedded component's blocks are copy-pasted
+into that component's template. This has a few interesting consequences.
```
Comment (weaverryan) [inline src/TwigComponent/doc/index.rst]: into the target component’s template

File: src/TwigComponent/doc/index.rst [hunk 7]
```diff
@@ -832,6 +832,140 @@ The ``with`` data is what's mounted on the component object.
...
+
+First of all, although `this` is referring to the component of the template into which an (embedded)
```
Comment (weaverryan) [inline src/TwigComponent/doc/index.rst]: First, although ``this`` refers to…

File: src/TwigComponent/doc/index.rst [hunk 8]
```diff
@@ -832,6 +832,140 @@ The ``with`` data is what's mounted on the component object.
...
+component is nested, *INSIDE* an embedded component it is referring to the *embedded* component,
+*NOT* the component into which it's being nested.
```
Comment (weaverryan) [inline src/TwigComponent/doc/index.rst]: The wording on this is really tricky: nested vs embedded. Maybe say less and let the code do the talking.

File: src/TwigComponent/doc/index.rst [hunk 9]
```diff
@@ -832,6 +832,140 @@ The ``with`` data is what's mounted on the component object.
...
+lower components. (That's also why `this` refers to the embedded, or "current" component)
+
+Finally, the most interesting thing is that, because the embedded components are literally
+resolved when rendering the nested component's template, you can even access the variables
+from THAT template as well. Again, it's as if the embedded component's blocks are being
+copy-pasted into the nested component's template.
+
+.. code-block:: twig
+
+    {# templates/successAlert.html.twig #}
```
Comment (weaverryan) [inline src/TwigComponent/doc/index.rst]: SuccessAlert.html.twig - I think that’s in a few places.

**Review by weaverryan — COMMENTED** [review]
Parts are a bit more technically worded than I’d like - and I dropped you an idea on slack, but this is very nice: you chose excellent examples to show off what’s possible. Thanks!

File: src/TwigComponent/doc/index.rst (hunk 4 above)
Comment (sneakyvv) [reply src/TwigComponent/doc/index.rst]: I guess you're right. Essentially, it was kinda explaining how blocks work, which is not the scope of this documentation.

**Comment by weaverryan** [comment]
Hey! I just did a quick proof and edit of the docs so far - a team effort to explain these important concepts on a simple way. I've tried to remove the wording "embedded component". Even I mix up "embedded" and "nested" in my mind. It's not as succinct, but saying "when you pass a block to a component" instead of "when you render an embedded component" I think is much more clear.

**Comment by sneakyvv** [comment]
> Hey! I just did a quick proof and edit of the docs so far - a team effort to explain these important concepts on a simple way. I've tried to remove the wording "embedded component". Even I mix up "embedded" and "nested" in my mind. It's not as succinct, but saying "when you pass a block to a component" instead of "when you render an embedded component" I think is much more clear.

Love it. Indeed I also always have to think twice to correctly separate nested vs embedded components. It took a while to get this terminology right in my head, but it will take perhaps equally long to stop referring to them as embedded components :-) 
Perhaps on a technical level, calling them that is actually still the most correct, especially seeing their similarity to the `embed` syntax, and even more so since the `ComponentNode` is in fact extending the `EmbedNode`. 🤓 
But again, I'm all for simplicity and even more for clarity in the documentation!

**Comment by weaverryan** [comment]
> Perhaps on a technical level, calling them that is actually still the most correct

I think so yes. So we may sometimes still use that term - but let's see if we can avoid it for the "outside" world :)

**Comment by sneakyvv** [comment]
@weaverryan As discussed on Slack, I also allowed references to outer blocks from with a regular (non-nested) component. 
Note that these blocks probably only make sense within an extending template, because otherwise the block would be rendered where it is defined ànd where it is referred as an outerblock. It even that might be desirable in some specific use cases, who knows 🤷

**Comment by weaverryan** [comment]
Thanks Bart!

**Pull request merged**